### PR TITLE
Optimize decorator helper size

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Special thanks to the ts team, the inspiration for the second commit came from their helpers, that is the most critical commit!
The size is reduced by ~20% from 4566 to 3694, and the difference after gzip may not be that huge, only ~10%. (Even without including the last commit to enable more aggressive mangle rules)
In order to facilitate review, this PR does not contain any incompatible changes, so it may not be optimal yet. :)

Updated: 
In order for the minifier to work better and the output to be more beautiful, I made a lot of code moves in the last two commits.
It might be easier to review commit-by-commit. :)
Changes in the first four commits: https://github.com/babel/babel/pull/16160/files/7156240ccad505e5fbb69bf78078e4d187c7c6f7

### CODE REVIEW

**Comment by liuxingbaoyu** [comment]
https://github.com/babel/babel/actions/runs/7127393069/job/19407228664
https://github.com/babel/babel/actions/runs/7137758148/job/19438225613
I suspect that these two CI freezes are related to this PR, which is very strange.

File: packages/babel-helpers/src/helpers/applyDecs2305.ts [hunk 1]
```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
...
+    throwUndefined?: boolean,
+  ) {
+    if (typeof fn !== "function") {
+      if (throwUndefined || fn !== void 0) {
+        throw new TypeError(
+          hint1 +
+            " must " +
+            (hint2 || "be") +
+            " a function" +
+            (throwUndefined ? "" : " or void 0"),
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2305.ts:229]:
```suggestion
            (throwUndefined ? "" : " or undefined"),
```

Most people never see `void 0` when writing JS code :)

File: packages/babel-helpers/src/helpers/applyDecs2305.ts [hunk 2]
```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
...
+            };
+            if (isField) {
+              set = function (target: any, v: any) {
+                target[name] = v;
+              };
+            }
+          } else if (kind === PROP_KIND.METHOD) {
+            // Assert: isPrivate is true.
+            get = function (_this: any) {
+              if (isPrivate) {
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2305.ts:364]: If `isPrivate` is true, why do we need to check it again?

File: packages/babel-helpers/src/helpers/applyDecs2305.ts [hunk 3]
```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
...
+          "setter",
+          "field",
+          "class",
+        ][kind] as any,
+
+        name: name,
+        metadata: metadata,
+      };
+
+      ctx.addInitializer = function (
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2305.ts:331]: Why is this not inside the object literal above?

File: packages/babel-helpers/src/helpers/applyDecs2305.ts [hunk 4]
```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
...
+                "set",
+                isPrivate && assertInstanceIfPrivate,
+              );
+            }
+          }
+
+          var access: DecoratorContextAccess = (ctx.access = {
+            has: isPrivate
+              ? // @ts-expect-error no thisArg
+                hasPrivateBrand.bind()
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2305.ts:389]: Why do we need to `.bind()`?

File: packages/babel-helpers/src/helpers/applyDecs2305.ts [hunk 5]
```diff
@@ -674,32 +189,407 @@ function defineMetadata(Class: any, metadata: any) {
...
+
+        if (
+          existingKind === true ||
+          (existingKind === PROP_KIND.GETTER && kind !== PROP_KIND.SETTER) ||
+          (existingKind === PROP_KIND.SETTER && kind !== PROP_KIND.GETTER)
+        ) {
+          throw new Error(
+            "Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " +
+              name,
+          );
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2305.ts:522]: Q: Do you know how this should work?

File: packages/babel-helpers/src/helpers/applyDecs2305.ts (hunk 4 above)
Comment (liuxingbaoyu) [reply packages/babel-helpers/src/helpers/applyDecs2305.ts:389]: I don't know, the code just moved. There's even a test to make sure `bind` is called.
image

File: packages/babel-helpers/src/helpers/applyDecs2305.ts (hunk 3 above)
Comment (liuxingbaoyu) [reply packages/babel-helpers/src/helpers/applyDecs2305.ts:331]: Oops, I forgot this when rebasing https://github.com/babel/babel/pull/16132.

Comment (liuxingbaoyu) [reply packages/babel-helpers/src/helpers/applyDecs2305.ts:331]: Oops, I forgot this when rebasing https://github.com/babel/babel/pull/16132.

File: packages/babel-helpers/src/helpers/applyDecs2305.ts (hunk 2 above)
Comment (liuxingbaoyu) [reply packages/babel-helpers/src/helpers/applyDecs2305.ts:364]: This should also move the old code and I will remove this check.

File: packages/babel-helpers/src/helpers/applyDecs2305.ts (hunk 5 above)
Comment (liuxingbaoyu) [reply packages/babel-helpers/src/helpers/applyDecs2305.ts:518]: I don't know, but this message is a bit too long for me, and I plan to modify or remove it later.

**Review by nicolo-ribaudo — APPROVED** [review]
Great!

**Pull request merged**